### ISSUE TITLE
Issues with Reproducing Evaluation Results

### ISSUE THREAD

**Issue by YunSeoHwan** [LiuZH-19/SongGen#13]
Hi, Zihan

I'm writing to share some results from my attempt to reproduce the evaluation using your provided official checkpoints. I've tested both mixed and interleaving modes. However, some metrics don't align with the figures reported in the paper.

Specifically, I've observed lower performance for FAD, CLAP, and PER compared to your reported numbers. Additionally, KL and SECE scores are higher than the published results, with the KL score notably close to zero, which appears unusual.

For the evaluation, I utilized the scripts mentioned in your paper (audioldm_eval, stable-audio-metric, and resemblyzer repro). The detailed evaluation code I've used is provided below for your reference:

Could you help identify possible reasons for these discrepancies or verify if there's any issue with the code? One potential concern I have is that I did not apply normalization to -14 dB LUFS. If a separate normalization step is necessary, could you please clarify how this should be performed?

Thank you very much for your assistance—I greatly appreciate your support.

Best regards,
```python
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "1" # Set GPU before torch import
os.environ["HF_HOME"] = "/nas2/musicgen/tmp"
os.environ["TRANSFORMERS_CACHE"] = "/nas2/musicgen/tmp"
import torch
import argparse
from audioldm_eval import EvaluationHelper
import json
import shutil
import glob
from resemblyzer import VoiceEncoder, preprocess_wav
import numpy as np
from tqdm import tqdm
import warnings

# Configure to ignore specific FutureWarning from the transformers library.
warnings.filterwarnings("ignore", category=FutureWarning, message="The input name `inputs` is deprecated.*")

# from clap_scorer import ClapScorer # No longer using previous clap_scorer
import sys
import pandas as pd
import transformers
try:
    from audiobox_aesthetics.infer import initialize_predictor
except ImportError:
    initialize_predictor = None
try:
    from transformers import pipeline
    from g2p_en import G2p
    import jiwer
except ImportError:
    pipeline, G2p, jiwer = None, None, None # Bug fix

# Add stable-audio-metrics path
# The library folder itself should be added to the path, not its parent.
metrics_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'stable-audio-metrics'))
if os.path.isdir(metrics_path):
    sys.path.append(metrics_path)
    from src.clap_score import clap_score # Function name changed
else:
    clap_score = None

# --- MERT Model Loading Patch Start ---
# Prevents audioldm_eval from directly loading a broken MERT model from HuggingFace,
# forcing it to use a locally saved, fixed version.
FIXED_MERT_PATH = "/nas2/musicgen/songgen/fixed_mert_v1_95m"
original_from_pretrained = transformers.AutoModel.from_pretrained

def patched_from_pretrained(pretrained_model_name_or_path, *model_args, **kwargs):
    # Intercept when audioldm_eval tries to load the MERT-95M model.
    if pretrained_model_name_or_path == "m-a-p/MERT-v1-95M":
        print(f"\nIntercepting model loading for '{pretrained_model_name_or_path}'...")
        print(f"Redirecting to local fixed model: '{FIXED_MERT_PATH}'")
        # Load from the local path instead.
        return original_from_pretrained(FIXED_MERT_PATH, *model_args, **kwargs)
    
    # For all other models, use the original function as is.
    return original_from_pretrained(pretrained_model_name_or_path, *model_args, **kwargs)

# Apply monkey patch
transformers.AutoModel.from_pretrained = patched_from_pretrained
# --- MERT Model Loading Patch End ---


def main(args):
    """
    Calculates FAD and KL metrics for generated and ground truth audio at given paths.
    Creates a temporary directory to match file lists for 'Paired mode' evaluation.
    """
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print(f"Using device: {device}")

    if not os.path.isdir(args.generated_path):
        print(f"Error: Generated audio path is not a directory or does not exist: {args.generated_path}")
        return
    if not os.path.isdir(args.gt_path):
        print(f"Error: Ground truth audio path is not a directory or does not exist: {args.gt_path}")
        return

    # Set up temporary GT directory for Paired-mode evaluation
    temp_gt_path = os.path.join(os.path.dirname(args.gt_path), "temp_gt_for_eval_mixed_2")
    
    try:
        # --- Load Text Metadata (adding lyrics) ---
        captions = {}
        lyrics_data = {}
        if os.path.exists(args.jsonl_path):
            print(f"\nLoading metadata for CLAP and PER score calculation: {args.jsonl_path}")
            with open(args.jsonl_path, 'r') as f:
                for line in f:
                    item = json.loads(line)
                    key = f"{item['fname']}.wav"
                    captions[key] = item['caption']
                    lyrics_data[key] = item.get('lyrics', '') # Handle case where lyrics might be missing
            print(f"Loaded {len(captions)} captions and {len(lyrics_data)} lyrics.")
        else:
            print(f"Warning: Metadata file ({args.jsonl_path}) not found. CLAP and PER scores cannot be calculated.")
            captions = None
            lyrics_data = None
        # --- End Load Text Metadata ---

        # Get list of generated files
        generated_files = os.listdir(args.generated_path)
        generated_basenames = {os.path.basename(f) for f in generated_files if f.endswith('.wav')}

        if not generated_basenames:
            print(f"Error: No .wav files found in generated audio directory '{args.generated_path}'.")
            return

        # Create temporary directory
        os.makedirs(temp_gt_path, exist_ok=True)
        print(f"Created temporary Ground-Truth directory: {temp_gt_path}")

        # Create symbolic links for matching GT files
        gt_files_linked = 0
        for basename in generated_basenames:
            src_gt_file = os.path.join(args.gt_path, basename)
            dst_gt_link = os.path.join(temp_gt_path, basename)
            if os.path.exists(src_gt_file):
                if os.path.lexists(dst_gt_link):
                    os.remove(dst_gt_link)
                os.symlink(src_gt_file, dst_gt_link)
                gt_files_linked += 1

        print(f"Created symbolic links for {gt_files_linked} Ground-Truth files matching generated files.")
        if gt_files_linked == 0:
            print("Error: No matching ground truth files found for generated files.")
            return

        # --- SECS Calculation Start ---
        print("\nStarting SECS (Speaker Encoder Cosine Similarity) calculation.")
        try:
            encoder = VoiceEncoder(device="cuda" if torch.cuda.is_available() else "cpu")
            similarity_scores = []

            for basename in tqdm(generated_basenames, desc="Calculating SECS"):
                try:
                    gen_path = os.path.join(args.generated_path, basename)
                    gt_path = os.path.join(temp_gt_path, basename)

                    wav_gen = preprocess_wav(gen_path)
                    wav_gt = preprocess_wav(gt_path)

                    embed_gen = encoder.embed_utterance(wav_gen)
                    embed_gt = encoder.embed_utterance(wav_gt)

                    # Calculate cosine similarity
                    similarity = np.dot(embed_gen, embed_gt) / (np.linalg.norm(embed_gen) * np.linalg.norm(embed_gt))
                    similarity_scores.append(similarity)
                except Exception as e:
                    print(f"Error calculating SECS for file '{basename}': {e}")
            
            secs_score = np.mean(similarity_scores) if similarity_scores else "N/A"
            print(f"SECS calculation complete. Average similarity: {secs_score}")
        except Exception as e:
            print(f"Error during Resemblyzer initialization or processing: {e}")
            secs_score = "N/A"
        # --- SECS Calculation End ---

        # --- CLAP Score Calculation Start ---
        if captions and clap_score:
            print("\nStarting CLAP Score calculation.")
            try:
                # Convert to id2text (dict) format required by clap_score function
                # Here, id is the basename of the file (without extension).
                id2text = {}
                for basename in generated_basenames:
                    # Now the keys in the captions dictionary and basename match exactly.
                    if basename in captions:
                        fname_without_ext = os.path.splitext(basename)[0]
                        id2text[fname_without_ext] = captions[basename]

                if id2text:
                    # Explicitly specify a compatible non-fusion model to prevent model loading errors.
                    clap_score_avg = clap_score(
                        id2text, 
                        args.generated_path,
                        clap_model='music_speech_audioset_epoch_15_esc_89.98.pt'
                    )
                    print(f"CLAP Score calculation complete. Average score: {clap_score_avg}")
                else:
                    print("Warning: No valid audio-caption pairs found for CLAP score calculation.")
                    clap_score_avg = "N/A"

            except Exception as e:
                print(f"Error during CLAP Score calculation: {e}")
                clap_score_avg = "N/A"
        else:
            if not clap_score:
                print("Warning: stable-audio-metrics library not found. CLAP score cannot be calculated.")
            clap_score_avg = "N/A"
        # --- CLAP Score Calculation End ---

        # --- Aesthetics Score (CE, CU, PC, PQ) Calculation Start ---
        if initialize_predictor:
            print("\nCalculating Aesthetics Scores (CE, CU, PC, PQ)...")
            try:
                # Correction: initialize_predictor does not take a device argument.
                predictor = initialize_predictor()
                
                # Prepare input data for prediction
                aes_input_data = [{"path": os.path.join(args.generated_path, basename)} for basename in generated_basenames]
                
                if aes_input_data:
                    # Manual batch processing to prevent OOM
                    batch_size = 16  # Adjust batch size appropriately to control memory usage
                    all_results = []
                    for i in tqdm(range(0, len(aes_input_data), batch_size), desc="Aesthetics Scores (batching)"):
                        batch = aes_input_data[i:i + batch_size]
                        batch_results = predictor.forward(batch)
                        all_results.extend(batch_results)

                    # Aggregate scores
                    ce_scores = [r['CE'] for r in all_results]
                    cu_scores = [r['CU'] for r in all_results]
                    pc_scores = [r['PC'] for r in all_results]
                    pq_scores = [r['PQ'] for r in all_results]
                    
                    avg_ce = np.mean(ce_scores)
                    avg_cu = np.mean(cu_scores)
                    avg_pc = np.mean(pc_scores)
                    avg_pq = np.mean(pq_scores)
                    
                    print("Aesthetics Scores calculation complete.")
                else:
                    avg_ce, avg_cu, avg_pc, avg_pq = ("N/A",) * 4
            except Exception as e:
                print(f"Aesthetics Scores calculation failed: {e}")
                avg_ce, avg_cu, avg_pc, avg_pq = ("N/A",) * 4
        else:
            print("Warning: audiobox-aesthetics library not found. Skipping aesthetic quality score calculation.")
            avg_ce, avg_cu, avg_pc, avg_pq = ("N/A",) * 4
        # --- Aesthetics Score Calculation End ---

        # --- Phoneme Error Rate (PER) Calculation Start ---
        if lyrics_data and pipeline and G2p and jiwer:
            print("\nCalculating Phoneme Error Rate (PER)...")
            try:
                # Check and download NLTK resources if needed.
                try:
                    import nltk
                    nltk.data.find('taggers/averaged_perceptron_tagger_eng')
                except ImportError:
                    print("NLTK is not installed.")
                except LookupError:
                    print("Downloading NLTK resource 'averaged_perceptron_tagger_eng'...")
                    nltk.download('averaged_perceptron_tagger_eng')
                    print("Download complete.")

                # 1. Load Distil-Whisper ASR pipeline
                print("Loading Distil-Whisper model for PER calculation...")
                asr_pipeline = pipeline(
                    "automatic-speech-recognition",
                    model="distil-whisper/distil-large-v3",
                    torch_dtype=torch.float16 if torch.cuda.is_available() else torch.float32,
                    device=device
                )
                # 2. Load G2P converter
                g2p = G2p()

                # 3. Perform ASR and calculate PER (changed to individual file processing for stability)
                per_scores = []
                # Prepare list of files to process for PER in advance
                files_to_process = [b for b in generated_basenames if b in lyrics_data and lyrics_data[b]]

                if files_to_process:
                    print(f"Transcribing {len(files_to_process)} audio files for PER...")
                    for basename in tqdm(files_to_process, desc="Calculating PER"):
                        try:
                            audio_path = os.path.join(args.generated_path, basename)
                            reference_lyrics = lyrics_data[basename]
                            
                            # Perform speech recognition for individual file
                            transcribed_text = asr_pipeline(audio_path, batch_size=1)['text']

                            # Phoneme conversion
                            ref_phonemes = " ".join(g2p(reference_lyrics))
                            hyp_phonemes = " ".join(g2p(transcribed_text))
                            
                            # Calculate PER
                            error = jiwer.wer(ref_phonemes, hyp_phonemes)
                            per_scores.append(error)
                        except Exception as e:
                            print(f"Warning: Could not calculate PER for '{basename}'. Reason: {e}")

                    avg_per = np.mean(per_scores) if per_scores else "N/A"
                    print(f"Phoneme Error Rate (PER) calculation complete. Average PER: {avg_per}")
                else:
                    print("Warning: No audio files with corresponding lyrics found for PER calculation.")
                    avg_per = "N/A"
            except Exception as e:
                print(f"PER calculation failed: {e}")
                avg_per = "N/A"
        else:
            print("Warning: Lyrics data or required libraries (transformers, g2p-en, jiwer) not found. Skipping PER calculation.")
            avg_per = "N/A"
        # --- Phoneme Error Rate (PER) Calculation End ---

        # Initialize EvaluationHelper instance
        evaluator = EvaluationHelper(args.sampling_rate, device)

        # Perform evaluation in Paired-mode
        print(f"\nStarting Paired-mode evaluation comparing '{args.generated_path}' and temporary GT folder '{temp_gt_path}'.")
        
        metrics = evaluator.main(
            args.generated_path,
            temp_gt_path  # Use temporary directory
        )

        # Output results
        fad_score = metrics.get("frechet_audio_distance", "N/A")
        kl_score = metrics.get("kullback_leibler_divergence_softmax", "N/A")

        print("\n--- Final Evaluation Summary ---")
        print(f"Frechet Audio Distance (FAD): {fad_score}")
        print(f"KL Divergence (KL): {kl_score}")
        # print(f"Speaker Encoder Cosine Similarity (SECS): {secs_score}")
        # print(f"CLAP Score: {clap_score_avg}")
        print(f"Content Enjoyment (CE): {avg_ce}")
        print(f"Content Usefulness (CU): {avg_cu}")
        print(f"Production Complexity (PC): {avg_pc}")
        print(f"Production Quality (PQ): {avg_pq}")
        print(f"Phoneme Error Rate (PER): {avg_per}")
        print("------------------------------\n")

        print("All Calculated Metrics:")
        # metrics['secs'] = secs_score
        # # Convert clap_score_avg to float if it's a torch.Tensor
        # metrics['clap_score'] = float(clap_score_avg) if isinstance(clap_score_avg, torch.Tensor) else clap_score_avg
        metrics['CE'] = avg_ce
        metrics['CU'] = avg_cu
        metrics['PC'] = avg_pc
        metrics['PQ'] = avg_pq
        metrics['PER'] = avg_per
        
        # Convert all numeric values to Python floats for JSON serialization
        serializable_metrics = {}
        for key, value in metrics.items():
            if isinstance(value, (np.floating, np.integer)):
                serializable_metrics[key] = float(value)
            elif isinstance(value, torch.Tensor):
                serializable_metrics[key] = value.item()
            else:
                serializable_metrics[key] = value

        print(json.dumps(serializable_metrics, indent=4))

    except Exception as e:
        print(f"An error occurred during evaluation: {e}")
    finally:
        # Clean up temporary directory
        if os.path.exists(temp_gt_path):
            print(f"Deleting temporary Ground-Truth directory: {temp_gt_path}")
            shutil.rmtree(temp_gt_path)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description="Evaluate generated audio using audioldm_eval.")
    parser.add_argument('--generated_path', type=str, default='/nas2/musicgen/data/MusicCaps_Test_Song/eval_inter_4/music', help='Directory containing generated audio files.')
    parser.add_argument('--gt_path', type=str, default='/nas2/musicgen/data/MusicCaps_Test_Song/audios/raw_audios', help='Directory containing ground truth audio files.')
    parser.add_argument('--sampling_rate', type=int, default=16000, help='Sampling rate of the audio files.')
    parser.add_argument('--jsonl_path', type=str, default='/nas2/musicgen/data/MusicCaps_Test_Song/metadata_abspath.jsonl', help='JSONL metadata file containing captions required for CLAP score calculation.')

    args = parser.parse_args()
    main(args)
```
